### PR TITLE
fix: fix socket leak in impersonated_credentials

### LINKED ISSUE #1122: impersonated_credentials.Credentials.sign_bytes leaks socket
_#### Environment details

  - OS: Ubuntu 20.04
  - Python version: 3.8
  - pip version: 22.2.2
  - `google-auth` version: 2.11.0

#### Steps to reproduce

  1. Create a service account save the email in a environment variable `SA_EMAIL`.
  2. Enable service account impersonation for your user (principal YOUR_EMAIL) with the role `iam.serviceAccountTokenCreator`. i.e:
```bash
gcloud iam service-accounts add-iam-policy-binding \
  --role=roles/iam.serviceAccountTokenCreator \
  --member=serviceAccount:${YOUR_EMAIL} ${SA_EMAIL}
```
  3. Install google-auth and requests into your Python environment
     `pip install google-auth requests`.
  4. Paste the following code into a test Python file (e.g test.py):
```python
import sys

import google.auth  # type: ignore
import google.auth.impersonated_credentials  # type: ignore

import unittest

class TestSignBlob(unittest.TestCase):

  def test_sign_blob(self):
    credentials, _ = google.auth.default()

    service_account_email = 'agentic-local-sa@agenticcorp.iam.gserviceaccount.com'

    signing_credentials = google.auth.impersonated_credentials.Credentials(
      source_credentials=credentials,
      target_principal=service_account_email,
      target_scopes=('https://www.googleapis.com/auth/devstorage.read_only',),
      lifetime=300)
    self.assertNotEqual(signing_credentials.sign_bytes(b'test'), b'')

if __name__ == '__main__':
  unittest.main()
```

  5. Run the test
```shell
python test.py
```

  6. Observe that a socket has been leaked on test tear down:
```
sign_leak.py:20: ResourceWarning: unclosed <ssl.SSLSocket fd=4, family=AddressFamily.AF_INET, type=SocketKind.SOCK_STREAM, proto=6, laddr=('172.23.11.59', 53774), raddr=('172.217.164.106', 443)>
  self.assertNotEqual(signing_credentials.sign_bytes(b'test'), b'')
ResourceWarning: Enable tracemalloc to get the object allocation traceback
```

It looks like the bug is here https://github.com/googleapis/google-auth-library-python/blob/bb5c9791c64e2472a90ba7191f79f4c5fedb2538/google/auth/impersonated_credentials.py#L289

The requests session object is created but it's never closed. It _should_ eventually be closed by the GC but really since it's no longer required it should be closed in this method.

### PR DESCRIPTION
**Pull request by stewartmiles**
impersonated_credentials.Credentials.sign_bytes() created
a session that wasn't closed leaking a socket. This fixes
the issue by always closing the requests session after
a signing request is complete.

Fixes #1122

### CODE REVIEW

**Comment by stewartmiles** [comment]
The failing presubmit looks like nothing to do with my change https://source.cloud.google.com/results/invocations/29f9a2fd-8d3a-4251-88f0-4598f6910f6b/targets/github%2Fgoogle-auth-library-python%2Fsystem_tests/tests

```
verify_refresh = <function _verify_refresh at 0x7fb0f2b48bd0>

    def test_application_default_credentials(verify_refresh):
        credentials, project_id = google.auth.default()

        if EXPECT_PROJECT_ID is not None:
            assert project_id is not None

>       verify_refresh(credentials)

system_tests_sync/test_default.py:28:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
system_tests_sync/conftest.py:122: in _verify_refresh
    credentials.refresh(http_request)
.nox/default_explicit_authorized_user-2-7/lib/python2.7/site-packages/google/oauth2/credentials.py:310: in refresh
    enable_reauth_refresh=self._enable_reauth_refresh,
.nox/default_explicit_authorized_user-2-7/lib/python2.7/site-packages/google/oauth2/reauth.py:347: in refresh_grant
    _client._handle_error_response(response_data)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

response_data = {'error': 'invalid_grant', 'error_description': 'Token has been expired or revoked.'}

    def _handle_error_response(response_data):
        """Translates an error response into an exception.

        Args:
            response_data (Mapping | str): The decoded response data.

        Raises:
            google.auth.exceptions.RefreshError: The errors contained in response_data.
        """
        if isinstance(response_data, six.string_types):
            raise exceptions.RefreshError(response_data)
        try:
            error_details = "{}: {}".format(
                response_data["error"], response_data.get("error_description")
            )
        # If no details could be extracted, use the response data.
        except (KeyError, ValueError):
            error_details = json.dumps(response_data)

>       raise exceptions.RefreshError(error_details, response_data)
E       RefreshError: ('invalid_grant: Token has been expired or revoked.', {u'error_description': u'Token has been expired or revoked.', u'error': u'invalid_grant'})

.nox/default_explicit_authorized_user-2-7/lib/python2.7/site-packages/google/oauth2/_client.py:62: RefreshError

```

**Review by clundin25 — APPROVED** [review]

**Comment by stewartmiles** [comment]
@clundin25 thanks, looks like everything is green now :) I don't have a merge button so I assume you'll merge this patch when you're ready?

**Comment by clundin25** [comment]
@stewartmiles Thanks for the contribution!

**Pull request merged**
